feature: json component serializer

kezz opened the pull request:
This PR adds a new module providing an implementation agnostic JSON serializer. The intention behind this PR is to allow platforms to swap between or use alternative implementations of a json serializer (see #839) whilst still allowing consumers of the platform to work regardless of the backing impl.

work that needs doing:
- [x] decide whether we need to bother with legacy stuff (legacy nbt serializer too?)
- [x] common tests for other serializers to use
- [x] common jmh benchmarks for other serializers to use
- [x] better error message for missing impl
- [x] docs for this (also in javadocs)
- [x] common builder for JSON serializers
- [x] finish moving text-serializer-gson-legacy-impl out to a json version, remove the old one from the bom, and stop publishing it with 4.15.0 (anyone still using it can just depend on an old version)
- ~~[ ] do we want a common flavour of the `Provider` services that normal serializers have, intended to be implemented by platforms?~~ not worth doing in this first revision

kezz committed:
feature(text-serializer-json): initial work on a generic json component serializer

zml2008 committed:
chore(text-serializer-json): Wire up shared tests between json impls

zml2008 committed:
chore(text-serializer-json): Shared benchmarks

kezz committed:
chore: Move tests into json serializer fixtures

zml2008 commented:
I think it's worthwhile to port the LegacyHoverEventSerializer and text-serializer-gson-legacy-impl over to `JsonComponentSerializer` -- it doesn't have any direct gson deps.

With options, I think it's worthwhile to have a common builder interface exposed as well -- since there are a few configurable elements.

I'm not sure what your goals were with the `Fallback` interface were -- I don't see it having any practical effect since you already `orElse` to the fallback without needing service discovery.

kezz committed:
chore: Make the Fallback interface optional

zml2008 committed:
chore(text-serializer-json): fix style errors and rebase issus

zml2008 committed:
chore(text-serializer-gson): Use shared serializer constants

zml2008 committed:
text-serializer-json: Bring the legacy hover serializer up

This allows common implemenattions to be used across serializer
implementations.

zml2008 committed:
feat(text-serializer-json): Add a builder interface in

zml2008 committed:
Json -> JSON

zml2008 committed:
fix style, migrate legacy hover serializer

zml2008 committed:
fix case in service declaration

zml2008 committed:
chore(text-serializer-json): Add a better instance unavailable error message

zml2008 committed:
chore(text-serializer-json): JD tweaks
